### DISCUSSION TITLE
How does relayer know where to get the packet?

### DISCUSSION THREAD

**Discussion by ImJeremyHe** [informalsystems/ibc-rs#183] [Q&A] [answered]
As far as I know, chain stores the `PacketCommitment` and its `ChannelQueryServer` provides `packet_commitments` to tell relayers that some packets should be delivered.  But how does the relayer know exactly what the packet was about?

**Comment by plafer** [comment, marked as answer]
By design, the relayer doesn't know what the packet is about, and doesn't need to. This is analogous to the IP protocol: you don't need to know what the data is about to route the packet to the destination.

Intuitively, all the relayer does is send the packet data to the destination chain (on the appropriate channel/port), along with a proof that the packet really was sent by the sending chain. The destination chain then convinces itself that the packet really was sent (by checking the proof), and gives the packet data to the appropriate IBC module (i.e. bound to the destination port), e.g. the ICS-20 transfer module. In this example, the transfer module is the component that knows how to interpret the packet data.

Hopefully that gives an intuition as to why the relayer doesn't need any information about what the packet data is in order to do its jobs of relaying. Does that fully answer your question though?

**Reply by ImJeremyHe** [reply]
Thanks for your reply.  I'm sorry, I didn't express myself clearly.

> Intuitively, all the relayer does is send the packet data to the destination chain (on the appropriate channel/port), along with a proof that the packet really was sent by the sending chain. 

Yes, relayer does not care about what the packet data is about. 
But I think there must be a place to store the packet binary data in the chain so that relayer will send it to the counter chain. It just stores the commitment of the packet, right?

**Reply by plafer** [reply]
> But I think there must be a place to store the packet binary data in the chain so that relayer will send it to the counter chain.

That's right; the packet binary data is stored in events (e.g. [SendPacket event](https://github.com/cosmos/ibc-rs/blob/8bd8235405da649166aeddfcb653f7339bffff81/crates/ibc/src/core/ics04_channel/events.rs#L488-L490)), which are sent back to tendermint in [deliverTx response](https://github.com/tendermint/tendermint/blob/v0.34.x/spec/abci/abci.md#delivertx-1). The relayer gets the binary packet data from there.

> It just stores the commitment of the packet, right?

We need to distinguish between what is stored "on-chain" (the packet commitments) and "off-chain" by the tendermint node (events and more, making these available to relayers).

**Reply by ImJeremyHe** [reply]
Thanks!

**Comment by ImJeremyHe** [comment]
Is this description accurate?
In cosmos-sdk chain, relayer should get the exact packet from the block and find the commitment to know that this chain has knew this packet.

**Reply by plafer** [reply]
That's right. The relayer gets the packet data from the event, and needs to fetch the proof that the packet commitment was indeed stored at the [expected location](https://github.com/cosmos/ibc/blob/main/spec/core/ics-024-host-requirements/README.md#path-space). [Here](https://github.com/informalsystems/hermes/blob/41aec7154e4fee51fba803d021c4541ef8902228/crates/relayer/src/link/relay_path.rs#L1343) is the code that does that.
